### PR TITLE
Separate factories + integration updates

### PR DESCRIPTION
**Pull request by mariusmihaic**
- Created separate factories for `SovereignNotifier` and `WsClient` in order to be easily integrated into `mx-chain-go`
- Refactor notifier subscriber to be compatible to `HeadersPool` from `mx-chain-go`
- Return error on `Close` w.r.t to [this PR](https://github.com/multiversx/mx-chain-core-go/pull/200)

### CODE REVIEW

File: process/interface.go [hunk 1]
```diff
@@ -18,16 +18,16 @@ type ShardCoordinator interface {
 	IsInterfaceNil() bool
 }
 
-// ExtendedHeaderHandler defines what a subscribed handler to SovereignNotifier should do
-type ExtendedHeaderHandler interface {
-	ReceivedExtendedHeader(header *block.ShardHeaderExtended)
+// HeaderSubscriber defines a subscribed to incoming headers
```
Comment (SebastianMarian) [inline process/interface.go]: a subscriber

File: process/notifier/sovereignNotifier.go [hunk 2]
```diff
@@ -101,7 +102,7 @@ func (notifier *sovereignNotifier) Notify(outportBlock *outport.OutportBlock) er
 		IncomingMiniBlocks: mbs,
 	}
 
-	notifier.notifyHandlers(extendedHeader)
+	notifier.notifyHandlers(outportBlock.BlockData.HeaderHash, extendedHeader)
```
Comment (SebastianMarian) [inline process/notifier/sovereignNotifier.go]: We need the hash of extendedHeader here so it must be calculated. outportBlock.BlockData.HeaderHash is the hash of shardHeader.

File: process/interface.go [hunk 3]
```diff
@@ -18,16 +18,16 @@ type ShardCoordinator interface {
...
+type HeaderSubscriber interface {
+	AddHeader(headerHash []byte, header data.HeaderHandler)
```
Comment (SebastianMarian) [inline process/interface.go:23]: ReceivedHeader ?

File: process/notifier/sovereignNotifier.go (hunk 2 above)
Comment (SebastianMarian) [reply process/notifier/sovereignNotifier.go]: Also we need to push the transactions included in the incoming mini blocks into sovereign node TxPool

File: process/interface.go (hunk 1 above)
Comment (sasurobert) [reply process/interface.go]: +1

File: process/interface.go (hunk 3 above)
Comment (sasurobert) [reply process/interface.go:23]: +1

File: process/interface.go (hunk 1 above)
Comment (mariusmihaic) [reply process/interface.go]: Right, refactored

File: process/notifier/sovereignNotifier.go (hunk 2 above)
Comment (mariusmihaic) [reply process/notifier/sovereignNotifier.go]: > We need the hash of extendedHeader here so it must be calculated. outportBlock.BlockData.HeaderHash is the hash of shardHeader. 

Right, missed it. Refactored as suggested + unit tests

> Also we need to push the transactions included in the incoming mini blocks into sovereign node TxPool

Very good point. Added task for it

File: process/interface.go (hunk 3 above)
Comment (mariusmihaic) [reply process/interface.go:23]: No, this should be compatible with `AddHeader` from `HeadersPool` in `mx-chain-go`

**Review by sasurobert — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**